tools/ldmp.c

```diff
@@ -414,8 +414,8 @@ dump_mbi(MEMORY_BASIC_INFORMATION *mbi)
            "State:             %08x %s\n"
            "Protect:           %08x %s\n"
            "Type:              %08x %s\n",
-           mbi->BaseAddress,
-           mbi->AllocationBase,
+           (uint)mbi->BaseAddress,
```

Comment:
None of the casts to uint seem like the right thing here: they're going to incorrectly truncate for 64-bit.  This whole file needs to be properly ported to 64-bit.  Today it is disabled from 64-bit builds.  #118 covers porting it.

Please see https://github.com/DynamoRIO/dynamorio/wiki/Code-Style: use PFX and related macros for printing pointer-sized variables.

Reply:
My assumption was that this code wasn’t used in 64 bit mode as the original code cannot work in 64 bit. In 64 bit the passed pointer is 8 bytes and %x will tell printf to get 4 bytes (sizeof uint) from the stack and print that, thus also do a truncation, and even worse the next argument that is printed will use the remaining 4 bytes of the passed pointer instead of the bytes of the following parameter.

Using the cast will truncate in 64 bit mode (when the address is above 4G), but will also ensures that the other passed parameter are printed correctly. In 32 bit mode it will work fine.

Now given the fact that it needs to work for 64 bit I shall rework the passed format string with PFX.
I would still need a cast to ptr_uint_t as PFX is defined as %08x (32bit) or %08lx (64bit). The MSVC14 compiler generates an error when passing a pointer to a printf function when the format specifier requires an unsigned int.

An alternative solution would be to define the macro PFX as “0x%p” or “%p” depending if the used C-runtime library prints pointers with or without a leading 0x. But I assume that this has already been tried in the past and doesn’t work on all supported platforms, which is a pity as it would not require the cast.